### DISCUSSION TITLE
[AOP] Why are roles favored over genuine aspects

### DISCUSSION THREAD

**Discussion by Opcyc-Ben** [dotnet/csharplang#8050] [Q&A] [answered]
TL;DR:
As someone deeply invested in C# for years, I pose this question with genuine curiosity:
Why do the language developers keep adding syntactic sugar around **specifically** static extensions, but never considering actual aspects?

The long version:
The addition of 'Roles' for C#13, as mentioned on the [Planned Language Feature Page](https://github.com/dotnet/roslyn/blob/main/docs/Language%20Feature%20Status.md#working-set), feels like yet another repackaging of extensions. The elevator pitch for roles (#5496) is to permit outsiders to add interfaces to existing classes, explicitly excluding instance members, in the name of lightweight typing.

So basically this:
```cs
role Bar : Foo, ISomething {
    // ... ISomething implementation ...
}
```
Instead of this:
```cs
class Bar : Foo, ISomething {
    // ... ISomething implementation ...
}
```
Or worst case this:
```cs
// Using the new, shiny syntax
sealed class Foo;

class Bar : ISomething {
    Foo foo;
    // ... ISomething implementation ...
}
```
In theory a new zero-cost abstraction.

The proposal seems reasonable, aiming to sidestep the abstraction overhead of class extension. It proposes a more concise means of bundling extension methods, enabling the tagging of objects of a specified base type at design-time to specific extension methods without introducing these methods to all instances of the extended base type:
```cs
class Foo;

/// <remarks>
/// Runtime type is <see cref="Foo"/>.
/// </remarks>
role Bar : Foo {
    // ... Bar specific methods ...
}

/// <remarks>
/// Runtime type is also <see cref="Foo"/>, same as <see cref="Bar"/>.
/// </remarks>
role Baz : Foo {
    // ... Baz specific methods ...
}
```
Design-time inheritance without run-time inheritance.

This would be achieved by introducing two new class types: `role` (see above) and `extension`, each used in place of `class`. Both undergo type-erasure, hence the requirement for no instance state in those types. Leaving out the `static` modifier on members seems like a fatal error, but ok.

For instance, roles would enable annotating a data object (e.g., JSON documents) based on its known content, providing insight into its structure and fields. Leveraging design-time types to confidently access these fields by name, assured of their existence.
```cs
public role Customer : DataObject {
    public string Name => this["Name"].AsString();
}
```

This approach circumvents the need to inadvertently import the same extension methods for other objects of the same base type that might coincidentally exist in the same file. Consider a scenario where a controller manages various untyped responses, some identified as 'Customer' objects and others not. In such cases, roles offer a means to annotate this information on **some** instances of the base type without incurring unnecessary run-time overhead.

This idea falls apart as soon as the object is passed around.

Currently, there's no way to check at run-time if e.g. a `DataObject` is or was a `Customer`.
Without additional allocations, converting/casting from a base type to its 'role' is problematic if it' was ever passed around or stored in (one of) its base form(s).

This would result in either a one-way only cast to its base type, with no reversibility, or a two-way cast with unchecked conversion to the deriving role. However, assuming all role members must adhere to extension rules, this at least wouldn't lead to any new errors. While technically feasible, it would defeat the purpose.

Ensuring type-safe casting and supporting type-checking operators at runtime would necessitate a witness object, as pointed out in the proposal's comments, rendering the zero-abstraction cost idea useless.

Implementing most interfaces also necessitates some form of instance data, and some interfaces even explicitly call for property implementations, which most of the time means instance data. E.g. IEnumerator isn't possible statically without dirty hacks, because of function re-entry.

While I agree, I do feel obligated to contribute an IEnumerable 'extension' which is good enough for most cases. The proposal specifically highlights a use case involving IEnumerable after all.
Here's one of my favorite C# debugging tools (read: toys) I wrote:
```cs
//TLS file
using System;
using System.Collections;
using System.Collections.Generic;
foreach(var i in 1..5)
    Console.WriteLine(i);

public static class RangeIteration {
    public static IEnumerator<int> GetEnumerator(this Range r)
        => r.Start.IsFromEnd || r.End.IsFromEnd
        ? throw new NotSupportedException()
        : new RangeEnumerator(r.Start.Value, r.End.Value);
    
    public class RangeEnumerator(int current, int end):IEnumerator<int> {
        object IEnumerator.Current => Current;
        public int Current {get; private set;}
        public bool MoveNext() => (Current = current++) <= end;
        public void Reset() => throw new NotImplementedException();
        public void Dispose() => GC.SuppressFinalize(this);
    }
}
```
[Sharplab.io link](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUPgAwAE+xAdAMID2ANmwKYDGCULMAM4BuCtTrEALKPIAzFsk4RuACwAUANwjIaUXTBqMGAVgCUFGpYkBONVFMyxAZgkA2OoRoAlCDADmnACSCJzIEHwCNADeFlb4LvTugQCiMKhgoeEKADywCAB8NADinAip6ZkICmoIKlCC3r4BNMjm5FYdNAC8hcgMAMoI2ggMgYIAYsgsYKkYNLi4LQyzoxNTMzAYsZ00APw0tVMA7jQwnCcAciwI/agADncKIRjJAB7cnHcRMGptO5YgU7nRr+TjlDJhKrINR9QbDBgANQgbFQnDgSxWSJRnAc2xoePiHhBAXBlWqeRo3FQyEUMAQ6IpnE2phAKTSEKyyFydMKMXa/xYwAAVjwEDQ2RVIQpmNTaWKejQmLKmQgZP86C4KUqaSrogFVTQ7sgoFoQjRBKVhABfPEdQnAFjsGgAWRYGk4F04rwQv26hTU2rl3Upyrp2Gwpho2S6NCZGDV/0J+Ek3k4Fp9kYVhxYJzOl2ugTAdy4GTpnBe70+31+CZ2SZTABF6o8Lb6FUUmAN7ka0xNYMioAAvTg1OqCXH8yw28hWoA==)
Despite its playful nature, this underscores the compiler's preference for name-matched functions over actual interface implementations in some scenarios.

I understand this doesn't work in most cases; however, I feel IEnumerable wouldn't currently warrant a new feature, two keywords, and a runtime overhaul.

The second mentioned use-case, an IComparable implementation, would warrant a new feature, and so would IEnumerator (not IEnumerable) honestly, due to current language limitations. 
However, I don't think this is the direction C# should take. There are already so many mildly overlapping features. I don't think roles would add much beyond extensions.

---
While the idea of 'roles' piques my interest, it feels like a **third** attempt (after extensions and interface default implementation) to avoid the diamond problem. The proposed feature bears resemblance to aspects of Aspect-Oriented Programming (AOP), which many have longed for in C#.

Traditional object-oriented programming encounters hurdles when trying to follow the DRY principle with classes without a common ancestor. That is to say: If two classes need the same functionality, and they are not related, you end up writing helper functions or libraries, exposing private members, or duplicating code.

AOP offers a remedy to this, by allowing objects to be composed of many fully formed, instantiable components, allowing code reusability across class-trees, though at the cost of an actual subclass, rather than a zero-cost abstraction. Also referred to as the 'Compositional Pattern'. Disambiguation is usually solved via casting if necessary.

I also believe optimizing away redundant internal-only classes shouldn't be the users/programmer's job. The compiler does a much better job at tree-shaking anyway. \#AgressiveInlining

Several third-party frameworks have implemented aspects in C#, indicating a demand for such functionality.
Among many more, there are:
* PostSharp and #metalama.
* Spring.NET (still "planned", by VMWare)
* My favorite: [NRoles](https://github.com/jordao76/nroles/blob/wiki/ProjectHome.md) (one person hobby project)

There are two reasons why I'm writing this today:
1. Relying on a single person or small team to keep a repo up-to-date with the actual behemoth C# compiler, Roslyn, would be madness. [xz](https://robmensching.com/blog/posts/2024/03/30/a-microcosm-of-the-interactions-in-open-source-projects/), [core-js](https://github.com/zloirock/core-js/blob/master/docs/2023-02-14-so-whats-next.md), etc. The burnout sets in eventually. Not 'If'. 'When'.
2. The CIL supports this today. Or... Erm... Doesn't forbid it.

---
In conclusion; I'm genuinely curious:
Why the reluctance to embrace AOP or similar approaches, opting instead for what feels like band-aid solutions?

Roles *especially* aren't a new proposal, the idea was brought up years prior, and discarded because too many questions were left unanswered. Why is this the direction the language designers chose to pursue today?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> Why is this the direction the language designers chose to pursue today?

Because it seems to elegantly provide solutions to many problem spaces we're interested in.   More so than aspects.  Also, no one has proposed a solution to aspects that doesn't come with serious problems (like being very expensive).  As such, we're going with the solution that addresses the scenarios we care about, that has a design that seems workable, and that doesn't go against major goals we have (perf, for example).

**Reply by Opcyc-Ben** [reply]
What do you mean by 'very expensive'. I am genuinely considering pouring time into this.
What perf goals would it need to hit?

**Reply by CyrusNajmabadi** [reply]
Alloc free for example.

**Reply by HaloFour** [reply]
I'd expect AOP, as compile-time instrumentation, to be completely free at runtime.  You'd only pay for the overhead of the aspect code itself.

**Reply by CyrusNajmabadi** [reply]
I'm a little confused on this bit then:

> Currently, there's no way to check at run-time if e.g. a DataObject is or was a Customer.

In order to be able to do a runtime check, we'd need a different runtime object passed around.  Which will result in an allocation.  How were you envisioning solving this?

**Reply by HaloFour** [reply]
Honestly, I'm not sure how AOP would apply in this case.  Aspects don't change the signature of their pointcuts, only their behavior.  Maybe OP is talking about something like bytecode transformers?

**Comment by Opcyc-Ben** [comment]
Honestly so would I... Conceptually AOP is nothing but stitching together 2+ base classes, then noting down which classes those were. And I'm not particularly deep in this, but I'd assume that Interface-To-Member mapping which already exists could be adapted for it. I'll try to give that a go. At least for personal use.


________________________________

**Reply by HaloFour** [reply]
> Conceptually AOP is nothing but stitching together 2+ base classes

That isn't any form of AOP with which I'm familiar.  AOP is aspects that can intercept pointcuts and apply custom advice, but it doesn't affect the signature of any types or their members.  You couldn't use it to make a class appear to implement an interface.  You could use byte code transformers to do that, but that's beyond what AOP typically provides.  It's also a bit beyond what a compiler would be expected to do.  The compiler couldn't take a class from one assembly and make it appear to implement an interface from another assembly.

IMO, I don't see where AOP addresses the problems that roles/extensions are attempting to address.  I definitely would like to see C# gain AOP capabilities, though, in that you can apply aspects to the code that you write and have the compiler inject the advice directly into the emitted assembly, kind of like an interceptor but at the callee.

**Reply by Opcyc-Ben** [reply]
> AOP is aspects that can intercept pointcuts and apply custom advice, but it doesn't affect the signature of any types or their members.

I took some time to reflect, gather my thoughts and ask my coworkers as well.

The interface extension part of the proposal, allowing Interfaces to be tagged onto existing classes is genuinely interesting new tech.

.Net has an experimental '[Interceptors](https://github.com/dotnet/roslyn/blob/main/docs/features/interceptors.md)' feature, which if it moves forward pretty much does vanilla Aspect things. It reminds me off [Harmony](https://harmony.pardeike.net/) and other run-time patching solutions, except working its magic during compilation if I understand it correctly.

It seems to me there is an opportunity to combine the two ideas.
A syntax to extend objects by all sorts of functionality, even operators, that always fully compiles away, because under the hood all functions are static anyway.

I'll mess around with that.

**Comment by hez2010** [comment]
Extensions (both implicit and explicit (or roles)), especially allowing to implement interfaces outside a type is the fundamental of type classes, which is the thing I want it to be in C# the most. This will lead us a safer (without the risk of downcast) and more extensible 
(can implement interfaces for existing types then we can use it with generics together) type system, without needs of subtyping, and importantly, costless and efficient. 

Spring-like AOP is neither statically verifiable nor efficient at performance, and it can be really expensive to use at runtime. It's convenient but I don't think run-time interception is a right direction for such purpose.

**Reply by HaloFour** [reply]
> Spring-like AOP is neither statically verifiable nor efficient at performance, and it can be really expensive to use at runtime. It's convenient but I don't think run-time interception is a right direction for such purpose.

AOP as a compiler feature wouldn't have any of those limitations or the runtime overhead that Spring proxy-based AOP has.

I do agree that type-classes or extension implementation would be quite useful and solve different problems than AOP would.

**Reply by ghostnguyen** [reply]
You could try this AOP library which is code generate, AOT compatible, no more expensive at runtime.

https://www.nuget.org/packages/InterfaceFillerCodeGen/
